feature/text_based.py

```diff
@@ -0,0 +1,724 @@
+        self._process_df(input_df, categories, method, cost_metric)
+
+        # Run multi-level set covering optimization
+        if method == "exact":
+            selected = self._select_multi_level_optimization()
+        # Run KMeans selection algorithm
+        elif method == "kmeans":
+            selected = self._select_kmeans()
+        # Run greedy heuristic
+        elif method == "greedy":
+            selected = self._select_greedy()
+        # Run random selection algorithm
+        elif method == "random":
+            selected = self._select_random()
+        else:
+            raise NotImplementedError(f"{method} has not been implemented.")
+
+        # Sort selection
+        selected.sort()
+
+        return selected
+
+    def _process_df(self, input_df: pd.DataFrame, categories: List[int], method: str, cost_metric: Optional[str]) \
+            -> NoReturn:
+        """
+
+        Parameters
+        ----------
+        input_df: pd.DataFrame
+            Input data frame with categories and features of content to select from.
+
+        categories: List[int]
+            List of columns in data that contains categories/labels to be covered.
+
+        method: str
+            The function need method to prevent performing featurization in
+             the case that optimization method is random(t or unicost) or greedy(unicost)
+
+        cost_metric: str
+            The function need cost metric to prevent performing featurization in
+             the case that cost metric is None or unicost (for random and greedy)
+
+        Returns
+        -------
+        No return
+       --------
+        * categories is the list of labels. code for labels with categories moved to utils.py
+        """
+        self.matrix = categories.to_numpy()
+        self._num_rows, self._num_cols = self.matrix.shape
+
+        # Content feature (with TextWiser)
+        if method == "random" and cost_metric in [None, "unicost"]:
+            pass
+        elif method == "greedy" and cost_metric == "unicost":
+            pass
+        else:
+            feature_column = self.featurization_method.fit_transform(input_df)
+            self.features = np.array([eval(l) if isinstance(l, str) else l for l in feature_column.tolist()])
+            assert (len(self.features) == self._num_cols), \
+                f"Process Data Error: features size ({len(self.features)}) " \
+                f"should match the number of columns ({self._num_cols})"
+
+        assert (self.matrix.ndim == 2), "Process Data Error: matrix should 2D"
+        if self.selection_size is not None:
+            assert (self.selection_size <= self._num_cols), "Process Data Error: selection_size cannot exceed num columns"
+
+
+
+    def _select_random(self) -> List[int]:# it returns selected column indices
+        # Set the seed
+        random.seed(self.seed)
+        best_selected = []
+        best_covered = 0
+
+        if self.selection_size is None:
```

Comment:
You can create a private method for the inner loop, as we will keep doing this operation for many other places where selection_size is None.. 
So the code will simplift to: 
```
selected_size = _calculate_selection_size() if selection_size is None else selection_size
```

Reply:
There is a more advanced version that we can do, but I am not sold on it (because it splits the logic)

you can move that one-liner to the constructor, so that whenver a None selection_size comes in, as the input, we do that operation once in the constructor and not worry about it here.. EXCEPT when the option is "exact"... 

Let's not go there for now, but I am just making the connection